### ISSUE TITLE
UnsupportedRDFormatException raised by HTTPRepositoryConnection when adding data from URL

### ISSUE THREAD

**Issue by manuelfiorelli** [eclipse/rdf4j#4451]
### Current Behavior

My application based on RDF4J 3.7.3 failed to import data from the following URL:
http://semanticturkey.uniroma2.it/ns/mdr

The root cause of the exception is indicated by the "cause" part of the stack trace (which I reduced for clarity).

``` 
        ...
        at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:386)
        at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:883)
        at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1034)
        at java.lang.Thread.run(Thread.java:748)
Caused by: org.eclipse.rdf4j.rio.UnsupportedRDFormatException: Did not recognise RDF format string text/turtle
        at org.eclipse.rdf4j.rio.Rio.lambda$unsupportedFormat$1(Rio.java:701)
        at java.util.Optional.orElseThrow(Optional.java:290)
        at org.eclipse.rdf4j.repository.http.HTTPRepositoryConnection.add(HTTPRepositoryConnection.java:425)
        at org.eclipse.rdf4j.repository.base.RepositoryConnectionWrapper.add(RepositoryConnectionWrapper.java:208)
        at org.eclipse.rdf4j.repository.RepositoryConnection.add(RepositoryConnection.java:857)
        ...
```

In my opinion, the problem boils down to the following [code](https://github.com/eclipse/rdf4j/blob/3.7.3/core/repository/http/src/main/java/org/eclipse/rdf4j/repository/http/HTTPRepositoryConnection.java#L424):
```
dataFormat = Rio.getParserFormatForMIMEType(mimeType)
                                  .orElse(Rio.getParserFormatForFileName(url.getPath()).orElseThrow(Rio.unsupportedFormat(mimeType)));
```
As you can see the `orElseThrow` is invoked on the nested `Rio.getParserFormatForFileName` rather than on the outer `getParserFormatForMIMEType`: consequently, the exception is thrown whenever the URL path does not end with a suitable extension even if the server is returning a MIME type.

### Expected Behavior

It should add the data without problems :-D

### Steps To Reproduce

Assuming you have imported the client libraries for RDF4J and you have a repository at the URL http://localhost:7200/repositories/TestGDB (I've used GDB free edition locally, but the actual URL and implementation should not matter).

```
import org.eclipse.rdf4j.model.util.Values;
import org.eclipse.rdf4j.repository.RepositoryConnection;
import org.eclipse.rdf4j.repository.http.HTTPRepository;

import java.io.IOException;
import java.net.URL;

class Scratch {
    public static void main(String[] args) throws IOException {
        HTTPRepository httpRepository = new HTTPRepository("http://localhost:7200/repositories/TestGDB");
        httpRepository.init();
        try {
            try (RepositoryConnection conn = httpRepository.getConnection()) {
                conn.add(new URL("http://semanticturkey.uniroma2.it/ns/mdr"), (String)null, null, Values.iri("http://semanticturkey.uniroma2.it/ns/mdr"));
            }
        } finally {
             httpRepository.shutDown();
        }
    }
}
```

This program fails with the following exception:

```
Exception in thread "main" org.eclipse.rdf4j.rio.UnsupportedRDFormatException: Did not recognise RDF format string text/turtle
	at org.eclipse.rdf4j.rio.Rio.lambda$unsupportedFormat$1(Rio.java:701)
	at java.util.Optional.orElseThrow(Optional.java:290)
	at org.eclipse.rdf4j.repository.http.HTTPRepositoryConnection.add(HTTPRepositoryConnection.java:425)
	at Scratch.main(scratch.java:15)
```

### Version

3.7.3

### Are you interested in contributing a solution yourself?

Yes

### Anything else?

The exactly same bug should also affect [4.3.0-M1](https://github.com/eclipse/rdf4j/blob/4.3.0-M1/core/repository/http/src/main/java/org/eclipse/rdf4j/repository/http/HTTPRepositoryConnection.java#L416)

I don't know whether you are going to make new releases for the 3.x branch. However, we really need a patch release for the 3.x series are we can't upgrade to v 4.x right now and the reported bug is really critical for us.

**Comment by manuelfiorelli** [comment]
I think that the same piece of code is also affected by two bugs:
* a NullPointerException is the server does not return a mime type
* redirects from HTTP to HTTPS are not followed: see #2828

As a workaround in my application, I might consider to subclass `RDFLoader`, so that when loading from a URL I can take the input stream and pass to the connection add method. Upstream, this might not be acceptable; a possible alternative could thus to add a method that instead of using an `RDFHandler` consumes the input stream directly.

**Comment by abrokenjester** [comment]
Confirmed the bug - looks like we overlooked the fact that in a Java optional, the `.orElse` _always_ gets executed. 

There are no further bug fix releases for 3.7 planned I'm afraid.

**Comment by manuelfiorelli** [comment]
Hi @abrokenjester after having filed the issue, I managed to work around the problem in our code base, so we don't have an urgent need for a patch release. Anyways, thanks for the availability. Indeed, we are stuck to RDF4J 3.x because are are bound to Javaa 8 for a series of "old" dependencies. However, we are going to address this in the coming months, and then we will upgrade RDF4J, as well.

Coming back to the issue, my workaround was to use `RDFLoader` to implement content-negotiation for free, and then use the resulting `InputStream` to add data to the `RepositoryConnection`. Technically, I had to create a subclass of `RDFLoader`, and then override the method taking an `InputStream` as argument (as you can see I pass `null` for the parameter `rdfHandler`, which is ignored in my method implementation.

```java
			new RDFLoader(new ParserConfig(), SimpleValueFactory.getInstance()){
				@Override
				public void load(InputStream in, String baseURI, RDFFormat dataFormat, RDFHandler rdfHandler) throws IOException, RDFParseException, RDFHandlerException {
					conn.add(in, baseURI, dataFormat, conn.getValueFactory().createIRI(baseURI));
				}
			}.load(new URL(sourceURL), baseURI2, rdfFormat, null);
``` 

I was wondering whether RDFLoader could be extended (or some of its functionality moved somewhere else) to make my solution cleaner, and thus suitable for inclusion upstream.
